### DISCUSSION TITLE
datagrid with  WPF MVVM + Reactiveui

### DISCUSSION THREAD

**Discussion by kakstrz** [reactiveui/ReactiveUI#2964] [Q&A]
Hello everyone :)
 I need to create a datagrid with a few DataGridTemplateColumn columns from WPF MVVM + Reactiveui. Can anyone tell me how it should be done correctly? I've been thinking about creating a sub-view for rows, but don't know how to do it in a good way.

**Comment by Hermansson** [comment]
Hi.

I'm attempting the same thing, and have sort of succeeded, but is not sure I'm using best practice.

In my view (ReactiveUserControl), I'm subscribing to the relevant data located in the view model:

```
//Contructor..

 DataContext = new MyViewModel();

 this.WhenActivated(disposables =>
{
.....
this.WhenAnyValue(v => v.ViewModel.Preview).Subscribe(preview => UpdateGrid(preview));
}
```

```
void UpdateGrid(List<string> data)
{
            return;
}
```

I intend to fill the data grid programmatically, I've placed a breakpoint in my "UpdateGrid" function to verify that the hook works.

My concern is that the "WhenAnyValue" call, or rathers Visual Studio 2022, is giving my a warning, "ViewModel may be NULL here". Not sure if the IDE is being too clever for it's own good or if I'm breaking best practice. As you can see, I'm creating my ViewModel before calling "WhenActivated", so I'm thinking of suppressing the warning, unless some kind soul here can enlighten me why that might be a bad idea, an present a better solution.

My view xaml contains no logic, other than a name I can reference using `this.FindControl<DataGrid>("PreviewGrid")`.
![image](https://user-images.githubusercontent.com/6293155/147980996-e9cab151-ced2-4ee7-a0a0-e50183b2f980.png)

**Reply by heronbpv** [reply]
You did nothing wrong per see, as the **ViewModel** property is defined as nullable. This is because, at view creation, there's no **ViewModel** set, at least if you're using the view location mechanism that RxUI provides.
What you could do is simply write the lambda as `v.ViewModel?.Preview`, to make the bind ignore the call when the property is null.

**Reply by Hermansson** [reply]
Hi. 
Thanks for confirming that I'm not coloring outside the lines, so to say.

Using the '?'  was my first instinct, but was slapped with this:
![image](https://user-images.githubusercontent.com/6293155/148120312-cc600741-ab1c-4b1a-8fc2-523a856ba874.png)

Anyway, for the OP's benefit, this is how I ended up populating the Grid. My Grid is currently for presenting only (no two way binding), but that will be reworked later since I plan to let the end user have the option of renaming the headers.

```
  List<string> headers = preview.Headers;
  List<string[]> rows = preview.Preview;

  PreviewGrid.Columns.Clear();
  for (int i = 0; i < headers.Count; i++)
      PreviewGrid.Columns.Add(new DataGridTextColumn { Header = headers[i], Binding = new Binding($"[{i}]") });

  PreviewGrid.Items = rows;
 
```

**Reply by glennawatson** [reply]
In the past I've done a builder wrapper around the DataGrid that takes in a Expression<Func<T>> similar to what we do in RxUI to get the name of the property. Be a good candidate for source generators too I guess.

**Reply by heronbpv** [reply]
@Hermansson That's what I get from not testing it in a editor. Yeah, what you can do in this case is use an observable pipeline on the viewmodel and only move forward if it's not null, like this:
`this.WhenAnyValue(v => v.ViewModel).WhereNotNull().Select(x => x.FilePreview).Skip(1).Subscriber(...)`

**Reply by Hermansson** [reply]
Ahh.. nice one, the observable pipeline i truly flexible once you get past the initial learning curve. Your snippet almost feels obvious when I read it out loud, but I would not have thought of it.

**Reply by Hermansson** [reply]
Above also helped me organize the code, setting up all behaviors in the constructor can become somewhat messy. The proposed approach would let me do the following (if there is a need).

```
 this.WhenAnyValue(v => v.ViewModel)
                .WhereNotNull()
                .Subscribe(observer => { 
                    InitBindings();
                    InitValidations();
                    InitCommands();
                    InitObservableSubscriptions();
                });
```
Edit: I had to switch from `WhereNotNull().Do(..)` to `WhereNotNull().Subscribe(..)`

**Reply by heronbpv** [reply]
> In the past I've done a builder wrapper around the DataGrid that takes in a Expression<Func> similar to what we do in RxUI to get the name of the property. Be a good candidate for source generators too I guess.

What about using bind converters? Would they be of any help in his case, to coerce the row values between the vm collection and whatever the Datagrid uses behind the scenes in WPF?
